File: image-processing-pipeline/camera_pipeline.py
from typing import List, Any
import numpy as np
import rawpy
from exiftool import ExifToolHelper # pip install PyExifTool # https://sylikc.github.io/pyexiftool/examples.html
import matplotlib.pyplot as plt

from cct_utils import interpolate_cst, xyz2cct
from colour_demosaicing import demosaicing_CFA_Bayer_Menon2007


def get_metadata(img_path:str) -> dict:

	def list2arr(cm):
		return np.reshape(np.asarray(cm, dtype='float64'), (3, 3))
	
	def int_list(val):
		return [int(v) for v in val]

	def float_list(val):
		return [float(v) for v in val]

	def str2list(val):
		return val.split(' ')

	tag_map = {
		  'EXIF:ActiveArea': 'active_area'
		, 'EXIF:DefaultCropSize': 'default_size'
		, 'MakerNotes:PerChannelBlackLevel': 'black_level'
		, 'EXIF:WhiteLevel': 'white_level'
		, 'EXIF:CFAPattern2': 'cfa_pattern'
		, 'EXIF:ISO': 'iso'
		, 'EXIF:ExposureTime': 'exposure_time'
		, 'EXIF:ApertureValue': 'aperture'
		, 'EXIF:FocalLength': 'focal_length'
		, 'EXIF:ColorMatrix1': 'cm1'
		, 'EXIF:ColorMatrix2': 'cm2'
	}

	with ExifToolHelper() as et:
		exif_d = et.get_tags(img_path, tags=list(tag_map.keys()))[0]
	
	metadata = {}
	metadata = {tag_map[k]:v for k, v in exif_d.items() if k in list(tag_map.keys())}

	metadata['active_area'] = int_list(str2list(metadata['active_area']))
	metadata['default_size'] = int_list(str2list(metadata['default_size']))[::-1]
	metadata['black_level'] = int_list(str2list(metadata['black_level']))
	metadata['cfa_pattern'] = int_list(str2list(metadata['cfa_pattern']))
	metadata['aperture'] = round(metadata['aperture'],2)

	cm1 = float_list(str2list(metadata['cm1']))
	cm2 = float_list(str2list(metadata['cm2']))

	metadata['cm1'] = list2arr(cm1)
	metadata['cm2'] = list2arr(cm2)

	return metadata


def raw_cropping(img:np.ndarray, active_area:List[int], default_size:List[int]) -> np.ndarray:
	"""
		Crops pixels exposed to light.

		Args:
			img (np.ndarray): the image to crop.
			active_area (list): active area specification [ top, left, bottom, right ]
			default_size (list): default area specification [ bottom, right ]

			example:
				H, W = img.shape
				active_area =  [ 0, 0, H, W ]
		
		Returns:
			the cropped image.
	"""
	y_origin = active_area[0]
	x_origin = active_area[1]

	height = default_size[0]
	width = default_size[1]
	return img[y_origin:y_origin + height, x_origin:x_origin + width]


def normalize(img:np.ndarray, black_level, white_level) -> np.ndarray:
	""" 
		Black light subtraction plus normalization.

		Args:
			img (np.ndarray): the image to normalize, this image must be in bayer form.
			black_level: specifies the smallest sensor reading value.
			white_level: specifies the highest sensor reading value.
		
		Returns:
			the clipped normalized image.
	"""
	black_level_mask = np.zeros(img.shape)

	idx2by2 = [[0, 0], [0, 1], [1, 0], [1, 1]]
	step2 = 2
	for i, idx in enumerate(idx2by2):
		black_level_mask[idx[0]::step2, idx[1]::step2] = black_level[i]

	img = img.astype(np.float64)
	img = ( img - black_level_mask ) / ( white_level - black_level_mask )
	img = np.clip(img, a_min=0.0, a_max=1.0)
	return img


def demosaic(img:np.ndarray, cfa_pattern:List[int]) -> np.ndarray:
	"""
		Demosaicing Based Onwavelet Analysis of the Luminance Component
		https://ieeexplore.ieee.org/abstract/document/4379122

		Args:
			img (np.ndarray): the image for demosaicing, this image must be in bayer form.
			cfa_pattern (list): specifies the order of the channels.

		Returns:
			The demosaiced RGB image.
	"""
	max_val = 65535
	img = (img * max_val).astype(dtype=np.uint16)

	cfa_pattern_str = "".join(["RGB"[i] for i in cfa_pattern])

	img = demosaicing_CFA_Bayer_Menon2007(img, pattern=cfa_pattern_str)
	img = img.astype(dtype=np.float64) / max_val

	return img


def find_opt_cm(wp:List[float], cm1:np.ndarray, cm2:np.ndarray) -> np.ndarray:
	"""
		Args:
			cm1 (np.ndarray): adobe DNG color matrix for standard illuminant A.
			cm2 (np.ndarray): adobe DNG color matrix for standard illuminant D65.
			wp (list): the wp to calculate the interpolated color matrix for, expressed in CIE's XYZ colorspace.
		
		Results:
			A typically normalized color matrix capable of transforming from XYZ_wp to RGB_wp.
	"""

	def wp_cm_norm(cm, wp):
		"""
			Typical normalization for color matrices.
			Color matrices are typically normalized so that the characterization illuminant (wp),
			expressed in CIE's XYZ color space, just saturates in the camera space.
		"""
		maxval = np.max(cm @ wp)
		return (1.0 / maxval) * cm

	wp = np.reshape(np.asarray(wp), (3, 1)) # wp should be a column vec.
	cm1 = wp_cm_norm(cm1, wp)
	cm2 = wp_cm_norm(cm2, wp)
	cm = interpolate_cst(cm1, cm2, xyz2cct(wp)) # CIE's illum. E cct ~= 5454
	return cm


def white_balance(img:np.ndarray, wp:List[float], xyz2cam:np.ndarray, cfa_pattern:List[int]) -> np.ndarray:
	"""
		Applies White Balance on a bayer image.

		1. maps scene illumination white point (wp) from XYZ to camera's RGB raw space.
		2. divides each channel by scene's wp, now expressed in camera's RGB raw space (RAW channel multipliers method).
		3. clips the result in the nominal range [0,1].

		Args:
			img (np.ndarray): the image to white balance, this image must be in bayer form.
			wp (list): the image white-point expressed in CIE XYZ_E color space.
			xyz2cam (np.ndarray): color matrix for converting between XYZ and raw RGB.
			cfa_pattern (list): specifies the order of the channels.
		
		Returns:
			the white balanced image in bayer form.

	"""
	wp = np.array(wp).reshape(3,1)
	as_shot_neutral = xyz2cam @ wp
	
	idx2by2 = [[0, 0], [0, 1], [1, 0], [1, 1]]
	step2 = 2
	white_balanced_image = np.zeros(img.shape)
	for i, idx in enumerate(idx2by2):
		idx_y = idx[0]
		idx_x = idx[1]
		white_balanced_image[idx_y::step2, idx_x::step2] = img[idx_y::step2, idx_x::step2] / as_shot_neutral[cfa_pattern[i]]

	white_balanced_image = np.clip(white_balanced_image, 0.0, 1.0)
	return white_balanced_image


def cam_to_linsrgb(img:np.ndarray, xyz2cam:np.ndarray) -> np.ndarray:
	"""
		cam_E -> XYZ_E -> XYZ_D65 -> linear sRGB_D65
		Args:
			img (np.ndarray): the image to apply color space transform on.
			xyz2cam (np.ndarray): color matrix for converting between XYZ and raw RGB.
	"""

	def row_cm_norm(cm:np.ndarray) -> np.ndarray:
		""" White in camera coords [1, 1, 1] should translate to white [1, 1, 1] in linear sRGB. """
		return cm / np.sum(cm, axis=1, keepdims=True)


	# Bradford adaptation.
	e2d65 = np.array([ [  0.9531874, -0.0265906,  0.0238731 ]
					  ,[ -0.0382467,  1.0288406,  0.0094060 ]
					  ,[  0.0026068, -0.0030332,  1.0892565 ]])

	# XYZ to sRGB, under D65.
	xyz2srgb = np.array([[ 3.2404542, -1.5371385, -0.4985314 ],
						 [-0.9692660,  1.8760108,  0.0415560 ],
						 [ 0.0556434, -0.2040259,  1.0572252 ]])

	cam2srgb = xyz2srgb @ e2d65 @ np.linalg.inv(xyz2cam)
	cam2srgb = row_cm_norm(cam2srgb)
	img = np.squeeze(cam2srgb[None, None, :, :] @ img[:, :, :, None])
	img = np.clip(img, 0.0, 1.0)
	return img


def linsrgb_to_xyz(img:np.ndarray) -> np.ndarray:
	
	# Bradford adaptation.
	d652e = np.array( [[ 1.0503, 0.0271, -0.0233 ]
					 , [ 0.0391, 0.9730, -0.0093 ]
					 , [-0.0024, 0.0026,  0.9181 ]])

	# sRGB to XYZ, under D65.
	srgb2xyz = np.array( [[ 0.4124564, 0.3575761, 0.1804375 ]
						 ,[ 0.2126729, 0.7151522, 0.0721750 ]
						 ,[ 0.0193339, 0.1191920, 0.9503041 ]])
	
	srgb2xyz = d652e @ srgb2xyz
	img = np.squeeze(srgb2xyz[None, None, :, :] @ img[:, :, :, None])
	img = np.clip(img, 0.0, 1.0)
	return img


def gamma_correction(img:np.ndarray) -> np.ndarray:
	img = img ** (1.0 / 2.2)
	return img


def xyz_to_srgb(img):
	""" XYZ_E -> XYZ_D65 -> sRGB_D65 """
	# Bradford adaptation method.
	e2d65 = np.array([ [0.9531874, -0.0265906,  0.0238731]
					  ,[-0.0382467,  1.0288406,  0.0094060]
					  ,[0.0026068, -0.0030332,  1.0892565]])

	# Convertion matrix from XYZ to sRGB, under D65.
	xyz2srgb = np.array([[3.2404542, -1.5371385, -0.4985314],
						 [-0.9692660, 1.8760108, 0.0415560],
						 [0.0556434, -0.2040259, 1.0572252]])

	M = xyz2srgb @ e2d65
	img = np.squeeze(M[None, None, :, :] @ img[:, :, :, None])
	img = np.clip(img, 0.0, 1.0)
	img = gamma_correction(img)
	return img


def single_image(img_path:str, debug=False) -> (np.ndarray, dict):

	img = rawpy.imread(img_path).raw_image.copy()
	meta = get_metadata(img_path)

	img = raw_cropping(img, meta['active_area'], meta['default_size'])
	img = normalize(img, meta['black_level'], meta['white_level'])

	wp_e = [1., 1., 1.] # E
	cm1 = meta['cm1'] # A
	cm2 = meta['cm2'] # D65
	xyz2cam = find_opt_cm(wp_e, cm1, cm2)

	cfa_p = meta['cfa_pattern']
	img = white_balance(img, wp_e, xyz2cam, cfa_p)
	img = demosaic(img, cfa_p)
	
	# The whole cam2srgb matrix has to be row normalized
	# that is why we first aquire the linear sRGB image and
	# we then get the XYZ.
	img = cam_to_linsrgb(img, xyz2cam)
	xyz = linsrgb_to_xyz(img)

	if debug:
		srgb = gamma_correction(img)
		plt.imshow((srgb * 255).astype('uint8'))
		plt.show(block=False)

		# Get a figure by converting acquired XYZ img to sRGB.
		plt.figure()
		plt.imshow((xyz_to_srgb(xyz) * 255).astype('uint8'))
		plt.show()
	
	return xyz, meta


if __name__ == '__main__':
	# img_path = "/home/ktzevel/Desktop/old_samples_and_algorithm/manual_dataset/moving_front/IMG_0032.dng"
	img_path = "/home/ktzevel/Desktop/DNG/104TRFLR/IMG_0046.dng"
	single_image(img_path)

File: image-processing-pipeline/camera_pipeline.m
    % --------------------------------------------------------
% Processing DNG images in Matlab
% Based on:
% 1. https://rcsumner.net/raw_guide/RAWguide.pdf
% 2. https://doi.org/10.1117/1.OE.59.11.110801
%
% --------------------------------------------------------

clear;
clc;

warning off MATLAB:tifflib:TIFFReadDirectory:libraryWarning
% filename = '/home/ktzevel/Desktop/old_samples_and_algorithm/manual_dataset/moving_front/IMG_0032.dng'; % Put file name here
% filename = '/home/ktzevel/Desktop/old_samples_and_algorithm/manual_dataset/traffic_light_R/IMG_0028.dng'; % Put file name here
filename = '/home/ktzevel/Desktop/DNG/104TRFLR/IMG_0046.dng';

% --------------------------------------------------------
% Fetching the raw image.
% --------------------------------------------------------
t = Tiff(filename,'r');
offsets = getTag(t,'SubIFD');
setSubDirectory(t, offsets(1));
raw = read(t); % Create variable ’raw’, the Bayer CFA data
close(t);

% Fetching metadata.
meta_info = imfinfo(filename);

% --------------------------------------------------------
% Crop raw image according to active_area and DefaultCrop tags.
% --------------------------------------------------------
x_origin = meta_info.SubIFDs{1}.ActiveArea(2)+1; % +1 due to MATLAB indexing
width = meta_info.SubIFDs{1}.DefaultCropSize(1);
y_origin = meta_info.SubIFDs{1}.ActiveArea(1)+1;
height = meta_info.SubIFDs{1}.DefaultCropSize(2);
raw = double(raw( y_origin:y_origin + height - 1, x_origin:x_origin + width - 1 ));

% --------------------------------------------------------
% Linearization
% --------------------------------------------------------
% Some manifactures use non-linear compress schemes for storing raw images (e.g. Nikon)
% Canon cameras do not do that.
if isfield(meta_info.SubIFDs{1},'LinearizationTable')
	ltab=meta_info.SubIFDs{1}.LinearizationTable;
	raw = ltab(raw+1);
end

% --------------------------------------------------------
% Normalization
% --------------------------------------------------------
black = meta_info.SubIFDs{1}.BlackLevel(1);
saturation = meta_info.SubIFDs{1}.WhiteLevel;
lin_bayer = ( raw - black ) / ( saturation - black );

% Clips the image in [0,1]
lin_bayer = max(0, min(lin_bayer, 1));

% --------------------------------------------------------
% White balance
% --------------------------------------------------------
% Uses camera's AWB values.
%wb_multipliers = (meta_info.AsShotNeutral).^-1;

% Equal-energy radiator, [1.000, 1.000, 1.000].
wp_e = whitepoint("e")';

% Color-matrices go from XYZ to camera color basis.
cm1 = meta_info.ColorMatrix1; % CIE's A
cm1 = reshape(cm1, 3, 3)'; % These are stored as 1x9 in a C row-wise manner.

cm2 = meta_info.ColorMatrix2; % CIE's D65
cm2 = reshape(cm2, 3, 3)';

% Typically normalize color-matrices s.t. chracterization illuminant just saturates
% in the camera space. 
cm1 = wp_cm_norm(cm1, wp_e);
cm2 = wp_cm_norm(cm2, wp_e);

cct_e = 5454;
xyz2cam = optimize_cm(cm1, cm2, cct_e); % optimizing for illuminant E.

wp_e = xyz2cam * wp_e; % converting it to camera space coordinates.
wb_multipliers = wp_e' .^ -1;

wb_multipliers = wb_multipliers / wb_multipliers(2);
mask = wbmask(size(lin_bayer, 1), size(lin_bayer, 2), wb_multipliers, 'rggb');
balanced_bayer = lin_bayer .* mask;

% Clips the image in [0,1]
balanced_bayer = max(0, min(balanced_bayer, 1));

% --------------------------------------------------------
% Demosaicing
% --------------------------------------------------------
temp = balanced_bayer / max(balanced_bayer(:)) * 65535;
dem = double(demosaic(uint16(temp),'rggb')) / 65535;

% --------------------------------------------------------
% Color space transform
% --------------------------------------------------------

e2d65 = [ 0.9531874 -0.0265906  0.0238731; ...
		 -0.0382467  1.0288406  0.0094060; ...
 		  0.0026068 -0.0030332  1.0892565 ];

xyz2srgb = [3.2404542 -1.5371385 -0.4985314; ...
			-0.9692660  1.8760108  0.0415560; ...
			0.0556434 -0.2040259  1.0572252];

cam2srgb = xyz2srgb * e2d65 / xyz2cam;

% row-normalization (taken from dcraw source code).
% A white point (1.0, 1.0, 1.0) in the camera space should result
% in a white point in the sRGB space.
cam2srgb = cam2srgb ./ repmat(sum(cam2srgb , 2), 1, 3);

lin_srgb = apply_cmatrix(dem, cam2srgb);
lin_srgb = max(0, min(lin_srgb, 1)); % Clips to [0,1]

% xyz = apply_cmatrix(lin_srgb, inv(e2d65) * inv(xyz2srgb));

% Gamma correction.
nl_srgb = lin_srgb .^ (1 / 2.2);


imshow(nl_srgb)

% f = figure;
% imshow(apply_cmatrix(xyz, xyz2srgb).^ (1 / 2.2))



% -------------- Function definitions follow: -------------- %

function cm = wp_cm_norm(cm, wp)

	maxval = max(cm * wp);
	cm = cm / maxval;
end

function cm = optimize_cm(cm1, cm2, temp)
	% Optimizes color matrix for illuminant with the specified cct temperature.

	cct1 = 2856;
	cct2 = 6504;

	if temp <= cct1
		g = 1.0;
	elseif temp >= cct2
		g = 0.0;
	else
		tempinv = 1.0 / temp;
		cct1inv = 1.0 / cct1;
		cct2inv = 1.0 / cct2;
		g = (tempinv - cct2inv) / (cct1inv - cct2inv);
	end

	if g >= 1.0
		cm = cm1;
		return
	end

	if  g <= 0.0
		cm = cm2;
		return
	end

	cm = g * cm1 + (1.0 - g) * cm2;
end


function colormask = wbmask(m, n, wbmults, align)
	% COLORMASK = wbmask(M,N,WBMULTS,ALIGN)
	%
	% Makes a white-balance multiplicative mask for an image of size m-by-n
	% with RGB while balance multipliers WBMULTS = [R_scale G_scale B_scale].
	% ALIGN is string indicating Bayer arrangement: ’rggb’,’gbrg’,’grbg’,’bggr’
	colormask = wbmults(2)*ones(m,n); %Initialize to all green values
	switch align
		case 'rggb'
			colormask(1:2:end, 1:2:end) = wbmults(1); %r
			colormask(2:2:end, 2:2:end) = wbmults(3); %b
		case 'bggr'
			colormask(2:2:end, 2:2:end) = wbmults(1); %r
			colormask(1:2:end, 1:2:end) = wbmults(3); %b
		case 'grbg'
			colormask(1:2:end, 2:2:end) = wbmults(1); %r
			colormask(1:2:end, 2:2:end) = wbmults(3); %b
		case 'gbrg'
			colormask(2:2:end, 1:2:end) = wbmults(1); %r
			colormask(1:2:end, 2:2:end) = wbmults(3); %b
	end
end

function corrected = apply_cmatrix(im, cm)
	% CORRECTED = apply_cmatrix(IM, CM)
	%
	% Applies CM to RGB input IM. Finds the appropriate weighting of the
	% old color planes to form the new color planes, equivalent to but much
	% more efficient than applying a matrix transformation to each pixel.
	if size(im, 3) ~= 3
		error('Apply CM to RGB image only.')
	end

	r = cm(1,1) * im(:,:,1) + cm(1,2) * im(:,:,2) + cm(1,3) * im(:,:,3);
	g = cm(2,1) * im(:,:,1) + cm(2,2) * im(:,:,2) + cm(2,3) * im(:,:,3);
	b = cm(3,1) * im(:,:,1) + cm(3,2) * im(:,:,2) + cm(3,3) * im(:,:,3);

	corrected = cat(3, r, g, b);
end